onnxruntime/core/common/spin_pause.cc

```diff
@@ -0,0 +1,86 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#if defined(_M_AMD64)
+#include <intrin.h>
+#endif
+
+#if defined(__x86_64__)
+#include <xmmintrin.h>
+#endif
+
+#if defined(_M_AMD64) || defined(__x86_64__)
+#if defined(__linux__)
+#include <x86intrin.h>
+#include <immintrin.h>
+#include <cstdlib>
+#elif defined(_WIN32)
+#include <Windows.h>
+#endif
+#include <cstdint>
+#include <string>
+#include "core/common/cpuid_info.h"
+#endif
+
+#include "core/common/spin_pause.h"
+
+namespace onnxruntime {
+
+#if defined(_M_AMD64) || defined(__x86_64__)
+static const bool tpause = CPUIDInfo::GetCPUIDInfo().HasTPAUSE();
+static const std::uint64_t spin_delay_cycles = 1000;
+
+// Check if TPAUSE Environment Variable is set
+static const bool tpause_enabled = []() -> bool {
+  bool enabled = false;
+  std::string env_value;
+
+#ifdef _WIN32
+  if (!tpause) return false;  // Skip if CPU doesn't support TPAUSE
+
+  char buffer[256];
+  if (GetEnvironmentVariableA("ORT_TPAUSE", buffer, 256) != 0) {
```

Comment:
I don't suggest using an environment variable for this, but for completeness, we do have a cross-platform helper function to read one:
https://github.com/microsoft/onnxruntime/blob/acb8ffd5df0d6c4c6b12610807d02e75ef9a70ed/onnxruntime/core/platform/env.h#L271-L275